agents/src/voice/generation.ts

```
@@ -0,0 +1,156 @@
+import type { AudioFrame, AudioSource } from '@livekit/rtc-node';
+import { randomUUID } from 'node:crypto';
+import type { ReadableStream } from 'stream/web';
+import type { ChatContext } from '../llm/chat_context.js';
+import { IdentityTransform } from '../stream/identity_transform.js';
+import { Future } from '../utils.js';
+import type { LLMNode, TTSNode } from './io.js';
+
+/* @internal */
+export class _LLMGenerationData {
+  generatedText: string = '';
+  id: string;
+
+  constructor(public readonly textStream: ReadableStream<string>) {
+    // TODO(shubhra): standardize id generation - same as python
+    this.id = randomUUID();
+  }
+}
+
+export function performLLMInference(
+  node: LLMNode,
+  chatCtx: ChatContext,
+  modelSettings: any, // TODO(shubhra): add type
+): [Promise<void>, _LLMGenerationData] {
+  const textStream = new IdentityTransform<string>();
+  const writer = textStream.writable.getWriter();
+  const data = new _LLMGenerationData(textStream.readable);
+
+  const inferenceTask = async () => {
+    const llmStream = await node(chatCtx, modelSettings);
+    if (llmStream === null) {
+      return;
+    }
+    try {
+      const reader = llmStream.getReader();
+      while (true) {
+        const { done, value: chunk } = await reader.read();
+        if (done) {
+          break;
+        }
+        if (typeof chunk === 'string') {
+          data.generatedText += chunk;
+          writer.write(chunk);
+          // TODO(shubhra): better way to check??
+        } else if ('choices' in chunk) {
+          const content = chunk.choices[0]?.delta.content;
+          if (!content) continue;
+          data.generatedText += content;
+          writer.write(content);
+        } else {
+          //TODO(shubhra): get access to logger instance
+          console.error('Unexpected chunk type:', chunk);
```

Right now we initiate the logger within a class ex. `private logger = log()`. Is there an issue with doing this in the beginning of a file so we don't have to pass it in to functions? This is what we do in python. Let me know what you think @lukasIO

---

there was a reason why were doing it this way, but I don't remember the details

---

Oh it's because you have to call `initializeLogger` somehow before all the imports run. https://github.com/livekit/agents-js/blob/236d2c62cc28d9a5c19e73e1b98709f243b91a77/agents/src/log.ts#L22

mildly annoying, we can just throw an error here for now.